org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/ProjectManager.java

```diff
@@ -59,32 +60,49 @@ public class ProjectManager {
 	@Inject
 	protected IExternalContentSupport externalContentSupport;
 
-	private IndexState indexState = new IndexState();
+	private IndexState persistedIndexState = new IndexState();
+
+	private IndexState dirtyIndexState;
 
 	private URI baseDir;
 
 	private Procedure2<? super URI, ? super Iterable<Issue>> issueAcceptor;
 
-	private Provider<Map<String, ResourceDescriptionsData>> indexProvider;
+	private Provider<Map<String, ResourceDescriptionsData>> persistedIndexProvider;
+
+	private Provider<Map<String, ResourceDescriptionsData>> dirtyIndexProvider;
 
 	private IExternalContentSupport.IExternalContentProvider openedDocumentsContentProvider;
 
-	private XtextResourceSet resourceSet;
+	private IExternalContentSupport.IExternalContentProvider dirtyDocumentsContentProvider;
+
+	private XtextResourceSet persistedResourceSet;
```

Comment:
Two resource-sets: Does this mean that the entire project will be in memory twice?

Reply:
It means that the resources loaded for dirty editors are loaded into a different resource set than the one for building the main index. So only as much of the project as you have loaded to support your dirty editors are in memory twice. However, worst case that could be the entire project, yes.